core/src/main/java/org/infinispan/interceptors/TxInterceptor.java

```diff
@@ -447,4 +314,148 @@ public long getCommits() {
+                       globalTransaction, alreadyCompleted, originatorMissing);
+         }
+         RollbackCommand rollback = commandsFactory.buildRollbackCommand(command.getGlobalTransaction());
+         try {
+            invokeNextInterceptor(ctx, rollback);
+         } finally {
+            RemoteTransaction remoteTx = ctx.getCacheTransaction();
+            remoteTx.markForRollback(true);
+            txTable.removeRemoteTransaction(globalTransaction);
+         }
+      }
+
+      if (originatorMissing && !completedSuccessfully) {
+         throw log.orphanTransactionRolledBack(globalTransaction);
+      }
+   }
+
+   private Object enlistReadAndInvokeNext(InvocationContext ctx, VisitableCommand command) throws Throwable {
+      enlistIfNeeded(ctx);
+      return invokeNextInterceptor(ctx, command);
+   }
+
+   private void enlistIfNeeded(InvocationContext ctx) throws SystemException {
+      if (shouldEnlist(ctx)) {
+         enlist((TxInvocationContext) ctx);
+      }
+   }
+
+   private Object enlistWriteAndInvokeNext(InvocationContext ctx, WriteCommand command) throws Throwable {
+      LocalTransaction localTransaction = null;
+      if (shouldEnlist(ctx)) {
+         localTransaction = enlist((TxInvocationContext) ctx);
+         boolean implicitWith1Pc = useOnePhaseForAutoCommitTx && localTransaction.isImplicitTransaction();
+         if (implicitWith1Pc) {
+            //in this situation we don't support concurrent updates so skip locking entirely
+            command.setFlags(Flag.SKIP_LOCKING);
+         }
+      }
+      Object rv;
+      try {
+         rv = invokeNextInterceptor(ctx, command);
+      } catch (Throwable throwable) {
+         // Don't mark the transaction for rollback if it's fail silent (i.e. putForExternalRead)
+         if (ctx.isOriginLocal() && ctx.isInTxScope() && !command.hasFlag(Flag.FAIL_SILENTLY)) {
+            TxInvocationContext txCtx = (TxInvocationContext) ctx;
+            txCtx.getTransaction().setRollbackOnly();
+         }
+         throw throwable;
+      }
+      if (localTransaction != null && command.isSuccessful()) {
+         localTransaction.addModification(command);
+      }
+      return rv;
+   }
+
+   private boolean isNotValid(int status) {
+      return status != Status.STATUS_ACTIVE
+            && status != Status.STATUS_PREPARING
+            && status != Status.STATUS_COMMITTING;
+   }
+
+   private void replayRemoteTransactionIfNeeded(RemoteTxInvocationContext ctx, int topologyId) throws Throwable {
+      // If a commit is received for a transaction that doesn't have its 'lookedUpEntries' populated
+      // we know for sure this transaction is 2PC and was received via state transfer but the preceding PrepareCommand
+      // was not received by local node because it was executed on the previous key owners. We need to re-prepare
+      // the transaction on local node to ensure its locks are acquired and lookedUpEntries is properly populated.
```

Comment:
Maybe something like "If the transaction has a lower topology id, it might not have wrapped all the local entries (e.g. because it was received via state transfer). We need to ..."